File: ImperatorToCK3/Source/Imperator/Genes/GenesDB.cpp
#include "GenesDB.h"
#include "Log.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"



Imperator::GenesDB::GenesDB(const std::string& thePath) {
	registerKeys();
	parseFile(thePath);
	clearRegisteredKeywords();
}


Imperator::GenesDB::GenesDB(std::istream& theStream) {
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}


void Imperator::GenesDB::registerKeys() {
	registerKeyword("accessory_genes", [this](std::istream& theStream) {
		accessoryGenes = AccessoryGenes(theStream);
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}

File: ImperatorToCK3/Imperator/Genes/GenesDB.cs
using commonItems;

namespace ImperatorToCK3.Imperator.Genes {
	public class GenesDB : Parser {
		public AccessoryGenes Genes { get; private set; } = new();

		public GenesDB() { }
		public GenesDB(string path) {
			RegisterKeys();
			ParseFile(path);
			ClearRegisteredRules();
		}
		public GenesDB(BufferedReader reader) {
			RegisterKeys();
			ParseStream(reader);
			ClearRegisteredRules();
		}
		private void RegisterKeys() {
			RegisterKeyword("accessory_genes", reader =>
				Genes = new AccessoryGenes(reader)
			);
			RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreItem);
		}
	}
}
